### ISSUE TITLE
object doesn't have an __enter__/__exit__ yet docs say it does

### ISSUE THREAD

**Issue by jamesbraza** [python/cpython#97699]
# Documentation

Here's the current docs in the Python data model on [`object.__enter__`](https://docs.python.org/3/reference/datamodel.html#object.__enter__):

> `object.__enter__(self)`

And [`object.__exit__`](https://docs.python.org/3/reference/datamodel.html#object.__exit__):

> `object.__exit__(self, exc_type, exc_value, traceback)`

However, with Python 3.10.3, we see neither of these methods belong to `object`:

```python
>>> object.__enter__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'object' has no attribute '__enter__'. Did you mean: '__ne__'?
>>> object.__exit__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'object' has no attribute '__exit__'. Did you mean: '__init__'?
```

I think the docs could be:
- More clear that `__enter__` and `__exit__` are not methods on `object`
- Mention where they come from (just being defined within your object to match the interface)

Thank you for your consideration!

**Comment by stevendaprano** [comment]
This is a misreading of the documentation. `object` itself, the base class of all objects, does not and never has had `__enter__` and `__exit__` methods.

The entire page uses "object" to refer to some generic object, e.g. the section on **Emulating numeric types** uses "object" to refer to something that emulates a numeric value. In this section, "object" refers to any arbitrary object whose type defines the two `__enter__` and `__exit__` methods, which makes it a context manager.

In hindsight we should have called the base class `Object`, to avoid ambiguity between arbitrary objects or instances, and the base class `object`.

In this case I suggest we change the documentation to use "instance" instead of "object" for the entire page, not just the context manager section.

**Comment by TeamSpen210** [comment]
Would it be a good idea to also add positional-only markers, to indicate that the names are conventions and don’t necessarily need to match the docs?

**Comment by AlexWaygood** [comment]
> In this case I suggest we change the documentation to use "instance" instead of "object" for the entire page, not just the context manager section.

That might require updating a _lot_ of links across the rest of CPython's documentation.

**Comment by stevendaprano** [comment]
> That might require updating a _lot_ of links across the rest of CPython's documentation.

It doesn't have to be done all at once. It could be done incrementally, a page at a time, as people notice or otherwise need to work on a page for something else.

The ambiguity between `object` and object is painful. We're probably stuck with it (unless we want to introduce `Object` as an alias and deprecate the lowercase version) but we can at least improve the docs as and when the pain leads to confusion, which it has here.

**Comment by AlexWaygood** [comment]
Yes, I agree that this would make the data model documentation less confusing if we referred to dunders as `instance.__await__`, `instance.__enter__`, etc.

My point is that a lot of links to these methods from other parts of the documentation would be broken if the data model docs were reworked in this way. For example, there are currently [six other docs files](https://github.com/python/cpython/search?q=object.__class_getitem__&type=) in the CPython repo that link to [the datamodel docs on `object.__class_getitem__`](https://docs.python.org/3/reference/datamodel.html#object.__class_getitem__). I think we might have to update all of those links if we started calling it `instance.__class_getitem__` instead.

**Comment by pochmann** [comment]
> a lot of links to these methods from other parts of the documentation might be broken

A lot of links from other parts of *the internet*, too, for example from Stack Overflow. Is there no way to have anchors for both or to redirect the old links?

**Comment by AlexWaygood** [comment]
> Is there no way to have anchors for both or to redirect the old links?

I'm not sure. @CAM-Gerlach might know more :)

**Comment by CAM-Gerlach** [comment]
> A lot of links from other parts of the internet, too, for example from Stack Overflow. Is there no way to have anchors for both or to redirect the old links?

We can redirect external links by creating a ref target label that matches the old anchor pointing to the relevant object. That won't for internal links using `:meth:`/`:func:`, though; those can be updated here, but it will break other docs sites linking to them via intersphinx (though it will generate a warning with `-n`, and visibly display as unlinked in the rendered docs).
